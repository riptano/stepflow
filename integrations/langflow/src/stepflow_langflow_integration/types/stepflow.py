"""Stepflow workflow and step type definitions."""

from dataclasses import dataclass, field
from typing import Dict, Any, List, Optional


@dataclass
class StepflowStep:
    """Represents a Stepflow workflow step."""
    
    id: str
    component: str
    input: Dict[str, Any]
    output_schema: Optional[Dict[str, Any]] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert step to dictionary representation."""
        result = {
            "id": self.id,
            "component": self.component,
            "input": self.input,
        }
        
        if self.output_schema:
            result["output_schema"] = self.output_schema
        
        return result


@dataclass 
class StepflowWorkflow:
    """Represents a Stepflow workflow."""
    
    name: str
    steps: List[StepflowStep]
    description: Optional[str] = None
    version: str = "1.0"
    schema: str = "https://stepflow.org/schemas/v1/flow.json"
    output: Optional[Dict[str, Any]] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert workflow to dictionary representation."""
        result = {
            "schema": self.schema,
            "name": self.name,
            "steps": [step.to_dict() for step in self.steps],
        }
        
        if self.description:
            result["description"] = self.description
        
        if self.version != "1.0":
            result["version"] = self.version
        
        if self.output:
            result["output"] = self.output
        
        return result