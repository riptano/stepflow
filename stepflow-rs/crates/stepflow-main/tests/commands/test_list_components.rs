// Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements.
// See the NOTICE file distributed with this work for additional information regarding copyright
// ownership.  The ASF licenses this file to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance with the License.  You may obtain a
// copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations under
// the License.

use super::stepflow;
use insta_cmd::assert_cmd_snapshot;
use std::fs;
use tempfile::TempDir;

#[test]
fn test_list_components_help() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("--log-level=error")
            .arg("list-components")
            .arg("--help")
    );
}

#[test]
fn test_list_components_default() {
    // Use the builtins config from tests/
    assert_cmd_snapshot!(
        stepflow()
            .arg("--log-level=error")
            .arg("list-components")
            .arg("--config=../../tests/stepflow-config.yml")
            .current_dir(env!("CARGO_MANIFEST_DIR"))
    );
}

#[test]
fn test_list_components_format_json() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("--log-level=error")
            .arg("list-components")
            .arg("--config=../../tests/stepflow-config.yml")
            .arg("--format=json")
            .current_dir(env!("CARGO_MANIFEST_DIR"))
    );
}

#[test]
fn test_list_components_format_yaml() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("--log-level=error")
            .arg("list-components")
            .arg("--config=../../tests/stepflow-config.yml")
            .arg("--format=yaml")
            .current_dir(env!("CARGO_MANIFEST_DIR"))
    );
}

#[test]
fn test_list_components_with_schemas() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("--log-level=error")
            .arg("list-components")
            .arg("--config=../../tests/stepflow-config.yml")
            .arg("--schemas=true")
            .current_dir(env!("CARGO_MANIFEST_DIR"))
    );
}

#[test]
fn test_list_components_json_no_schemas() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("--log-level=error")
            .arg("list-components")
            .arg("--config=../../tests/stepflow-config.yml")
            .arg("--format=json")
            .arg("--schemas=false")
            .current_dir(env!("CARGO_MANIFEST_DIR"))
    );
}

#[test]
fn test_list_components_custom_config() {
    let temp_dir = TempDir::new().expect("Failed to create temp directory");
    let config_path = temp_dir.path().join("custom-config.yml");

    // Create a custom config with builtin components
    let config_content = r#"
working_directory: .
plugins:
  - name: test-builtins
    type: builtin
"#;

    fs::write(&config_path, config_content).expect("Failed to write config file");

    assert_cmd_snapshot!(
        stepflow()
            .arg("--log-level=error")
            .arg("list-components")
            .arg(format!("--config={}", config_path.display()))
            .current_dir(env!("CARGO_MANIFEST_DIR"))
    );
}

#[test]
fn test_list_components_nonexistent_config() {
    // Use a fixed path that doesn't exist for consistent snapshots
    assert_cmd_snapshot!(
        stepflow()
            .arg("--log-level=error")
            .arg("list-components")
            .arg("--config=/tmp/nonexistent-stepflow-config.yml")
            .current_dir(env!("CARGO_MANIFEST_DIR"))
    );
}
