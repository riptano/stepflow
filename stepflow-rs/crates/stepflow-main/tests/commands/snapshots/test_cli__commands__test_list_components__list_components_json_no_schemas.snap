---
source: crates/stepflow-main/tests/commands/test_list_components.rs
info:
  program: stepflow
  args:
    - "--log-file=/dev/null"
    - "--omit-stack-trace"
    - "--log-level=error"
    - list-components
    - "--config=tests/stepflow-config.yml"
    - "--format=json"
    - "--schemas=false"
---
success: true
exit_code: 0
----- stdout -----
{
  "components": [
    {
      "component": "/create_messages",
      "description": "Create a chat message list from system instructions and user prompt",
      "plugin": "builtin",
      "routes": [
        {
          "path_pattern": "/builtin/{*component}",
          "resolved_path": "/builtin/create_messages",
          "conditions": [],
          "is_conditional": false
        }
      ]
    },
    {
      "component": "/eval",
      "description": "Execute a nested workflow with given input and return the result",
      "plugin": "builtin",
      "routes": [
        {
          "path_pattern": "/builtin/{*component}",
          "resolved_path": "/builtin/eval",
          "conditions": [],
          "is_conditional": false
        }
      ]
    },
    {
      "component": "/get_blob",
      "description": "Retrieve JSON data from a blob using its ID",
      "plugin": "builtin",
      "routes": [
        {
          "path_pattern": "/builtin/{*component}",
          "resolved_path": "/builtin/get_blob",
          "conditions": [],
          "is_conditional": false
        }
      ]
    },
    {
      "component": "/iterate",
      "description": "Iteratively apply a workflow until it returns a result instead of next",
      "plugin": "builtin",
      "routes": [
        {
          "path_pattern": "/builtin/{*component}",
          "resolved_path": "/builtin/iterate",
          "conditions": [],
          "is_conditional": false
        }
      ]
    },
    {
      "component": "/load_file",
      "description": "Load and parse a file (JSON, YAML, or text) from the filesystem",
      "plugin": "builtin",
      "routes": [
        {
          "path_pattern": "/builtin/{*component}",
          "resolved_path": "/builtin/load_file",
          "conditions": [],
          "is_conditional": false
        }
      ]
    },
    {
      "component": "/map",
      "description": "Apply a workflow to each item in a list and collect the results",
      "plugin": "builtin",
      "routes": [
        {
          "path_pattern": "/builtin/{*component}",
          "resolved_path": "/builtin/map",
          "conditions": [],
          "is_conditional": false
        }
      ]
    },
    {
      "component": "/openai",
      "description": "Send messages to OpenAI's chat completion API and get a response",
      "plugin": "builtin",
      "routes": [
        {
          "path_pattern": "/builtin/{*component}",
          "resolved_path": "/builtin/openai",
          "conditions": [],
          "is_conditional": false
        }
      ]
    },
    {
      "component": "/put_blob",
      "description": "Store JSON data as a blob and return its content-addressable ID",
      "plugin": "builtin",
      "routes": [
        {
          "path_pattern": "/builtin/{*component}",
          "resolved_path": "/builtin/put_blob",
          "conditions": [],
          "is_conditional": false
        }
      ]
    }
  ]
}

----- stderr -----
